File: day14_part1_2022.js
const fs = require('fs');

const input = fs.readFileSync('input.txt', 'utf8').trim().split('\n');

const grid = new Map();
input.forEach(line => {
    const points = line.split(' -> ').map(point => point.split(',').map(Number));
    for (let i = 0; i < points.length - 1; i++) {
        const [x1, y1] = points[i];
        const [x2, y2] = points[i + 1];
        if (x1 === x2) {
            for (let y = Math.min(y1, y2); y <= Math.max(y1, y2); y++) {
                grid.set(`${x1},${y}`, true);
            }
        } else {
            for (let x = Math.min(x1, x2); x <= Math.max(x1, x2); x++) {
                grid.set(`${x},${y1}`, true);
            }
        }
    }
});

console.log(fill(grid));

function fill(grid) {
    const floor = bounds(grid).max.y + 1;
    let sands = 0;
    let firstFloorTouch = 0;
    let full = false;
    while (!full) {
        if (grid.has('500,0')) {
            full = true;
        } else {
            let sand = [500, 0];
            let settled = false;
            while (!settled) {
                const [x, y] = sand;
                if (y === floor) {
                    if (firstFloorTouch === 0) {
                        firstFloorTouch = sands;
                    }
                    grid.set(`${x},${y}`, true);
                    break;
                }
                let nextSand = null;
                for (const [dx, dy] of [[0, 1], [-1, 1], [1, 1]]) {
                    const nx = x + dx;
                    const ny = y + dy;
                    if (!grid.has(`${nx},${ny}`)) {
                        nextSand = [nx, ny];
                        break;
                    }
                }
                if (nextSand) {
                    sand = nextSand;
                } else {
                    grid.set(`${x},${y}`, true);
                    settled = true;
                }
            }
            sands++;
        }
    }
    return firstFloorTouch;
}

function bounds(grid) {
    const points = Array.from(grid.keys()).map(point => point.split(',').map(Number));
    return boundss(points);
}

function boundss(points) {
    let minX = Infinity;
    let minY = Infinity;
    let maxX = -Infinity;
    let maxY = -Infinity;
    points.forEach(([x, y]) => {
        minX = Math.min(minX, x);
        minY = Math.min(minY, y);
        maxX = Math.max(maxX, x);
        maxY = Math.max(maxY, y);
    });
    return { min: { x: minX, y: minY }, max: { x: maxX, y: maxY } };
}

File: cpp/day14_part1_2022.cpp

#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <sstream>
#include <map>
#include <algorithm>

struct Point {
    int x, y;

    Point(int x_val = 0, int y_val = 0) : x(x_val), y(y_val) {}

    Point add(const Point& other) const {
        return Point(x + other.x, y + other.y);
    }

    bool operator<(const Point& other) const {
        if (y != other.y) {
            return y < other.y;
        }
        return x < other.x;
    }
};

std::pair<Point, Point> bounds(const std::map<Point, bool>& grid) {
    int min_x = 2000000000, min_y = 2000000000;
    int max_x = -2000000000, max_y = -2000000000;

    for (const auto& pair : grid) {
        const Point& p = pair.first;
        min_x = std::min(min_x, p.x);
        min_y = std::min(min_y, p.y);
        max_x = std::max(max_x, p.x);
        max_y = std::max(max_y, p.y);
    }
    return {Point(min_x, min_y), Point(max_x, max_y)};
}

int fill(std::map<Point, bool>& grid) {
    Point min_coords, max_coords;
    std::tie(min_coords, max_coords) = bounds(grid);
    
    int floor = max_coords.y + 1;

    int sands = 0;
    int first_floor_touch = 0;

    std::vector<Point> directions = {Point(0, 1), Point(-1, 1), Point(1, 1)};

    while (true) {
        Point sand = Point(500, 0);

        while (true) {
            if (sand.y == floor - 1) {
                if (first_floor_touch == 0) {
                    first_floor_touch = sands;
                }
                grid[sand] = true;
                break;
            }

            bool moved = false;
            for (const auto& d : directions) {
                Point new_sand = sand.add(d);
                if (grid.find(new_sand) == grid.end()) {
                    sand = new_sand;
                    moved = true;
                    break;
                }
            }

            if (!moved) {
                grid[sand] = true;
                break;
            }
        }
        
        sands++;

        if (sand.y == 0) {
            return first_floor_touch;
        }
    }
}

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);

    std::map<Point, bool> grid;

    std::ifstream inputFile("input.txt");
    std::string line;

    while (std::getline(inputFile, line)) {
        size_t pos = 0;
        while ((pos = line.find(" -> ", pos)) != std::string::npos) {
            line.replace(pos, 4, " ");
            pos += 1; 
        }

        std::vector<Point> points;
        std::istringstream ss(line);
        int x_val, y_val;
        char comma;
        while (ss >> x_val >> comma >> y_val) {
            points.emplace_back(x_val, y_val);
        }

        for (size_t i = 0; i < points.size() - 1; ++i) {
            Point p1 = points[i];
            Point p2 = points[i+1];

            if (p1.x == p2.x) {
                for (int y_fill = std::min(p1.y, p2.y); y_fill <= std::max(p1.y, p2.y); ++y_fill) {
                    grid[Point(p1.x, y_fill)] = true;
                }
            } else {
                for (int x_fill = std::min(p1.x, p2.x); x_fill <= std::max(p1.x, p2.x); ++x_fill) {
                    grid[Point(x_fill, p1.y)] = true;
                }
            }
        }
    }

    std::cout << fill(grid) << std::endl;

    return 0;
}
